File: Java/Piranha.cpp
#include "Piranha.hpp"
#include "Aquarium.hpp"

Piranha :: Piranha(){
	MAX_STARVING = 500;
	MAX_STARVATION_PERIOD = 200;
	MAX_MOVE_TIME = 70;
	speed  = 2;
	latest_food = 0;
	right_direct = false;
	degrees = 30;
	radius = 20;
	jenis = "Piranha";
	setStarvationPeriod(MAX_STARVATION_PERIOD);
	setStarving(MAX_STARVING);
	setMoveTime(MAX_MOVE_TIME);
}

void Piranha::produce(Aquarium& aq){
	Koin *koin = new Koin(getLast(),getPosisi());
    aq.AddObject(koin);
}

void Piranha::eat(Aquarium& aq){
	bool existFood = false;
	bool eat_food = false;
	double rad_min = 1e7;
  	int i=0,idx;
  	Point idx_piranha = getPosisi();
  	Point point_kejar;
  	while (i < aq.getListIkan().totalElmt()){
	    if (aq.getListIkan().get(i)->getJenis() == "Guppy"){
	    	existFood = true;
	    	Point point_makanan = aq.getListIkan().get(i)->getPosisi();
	    	double jarak = idx_piranha.hitungJarak(point_makanan);
	    	if(jarak < rad_min){
	    		rad_min = jarak;
	    		idx = i;
	    		point_kejar = point_makanan;
	    	}
	    } 
	    i++;
  	}
  	// Cek mau dimakan atau dikejar
  	if(existFood){
  		if(rad_min <= getRadius()){
  			eat_food = true;
  		}
  		else{
  			swimto(point_kejar, getSpeed());
			setStarving(getStarving()-1);
  		}
  	}

	// Makan guppy
	if(eat_food){
		//Update hungry time
		setStarvationPeriod(MAX_STARVATION_PERIOD);
		setStarving(MAX_STARVING);
		//ilangin guppy
		setLast(aq.getListIkan().get(idx)->getHarga()); //untuk makan
		aq.RemoveObject(aq.getListIkan().get(idx));
		produce(aq);
	}
	else{
		setStarving(getStarving()-1);
	}
}
int Piranha :: getProduceTime(){return 0;}
int Piranha :: getMaxProduceTime(){return 0;}
void Piranha :: setProduceTime(int x){}

int Piranha :: getLast(){
	return latest_food;
}
void Piranha :: setLast(int x){
	latest_food = x;
}

File: Java/Piranha.java
public class Piranha extends Fish {

  private int latestFood;

  public Piranha() {
    maxStarving = 500;
    maxStarvationPeriod = 200;
    maxMoveTime = 70;

    degree = 30;
    jenis = "Piranha";
    latestFood = 0;
    radius = 15;
    id = counter++;
    rightDirect = false;
    setStarvationPeriod(maxStarvationPeriod);
    setStarving(maxStarving);
    setMoveTime(maxMoveTime);
    speed = 2;
  }

  public boolean chaseFood(double radMin,Point pointKejar){
    if (radMin <= getRadius()) {
      return true;
    } else {
      swimto(pointKejar, getSpeed());
      setStarving(getStarving() - 1);
      return false;
    }
  }

  public void biteFish(Aquarium aquarium,int idx){
    setStarvationPeriod(maxStarvationPeriod);
    setStarving(maxStarving);

    Guppy temp = (Guppy) aquarium.getListIkan().get(idx);
    setLast(temp.getHarga() * (temp.getGrowthStep() + 1)); //untuk makan
    aquarium.removeObject(aquarium.getListIkan().get(idx));
    this.produce(aquarium);
  }

  public void eat(Aquarium aquarium) {
    boolean existFood = false;
    boolean eatFood = false;
    double radMin = 1e7;
    int i = 0;
    int idx = 0;
    Point idxPiranha = getPosisi();
    Point pointKejar = new Point();

    // check food exist
    while (i < aquarium.getListIkan().totalElmt()) {
      if (aquarium.getListIkan().get(i).getJenis().equals("Guppy")) {
        existFood = true;
        Point pointMakanan = aquarium.getListIkan().get(i).getPosisi();
        double jarak = idxPiranha.hitungJarak(pointMakanan);
        if (jarak < radMin) {
          radMin = jarak;
          idx = i;
          pointKejar = pointMakanan;
        }
      }
      i++;
    }

    // check walk to food or eat it
    if (existFood) {
      eatFood = chaseFood(radMin,pointKejar);
    }

    // eat guppy
    if (eatFood) {
      System.out.println("SEE ME");
      biteFish(aquarium, idx);
    } else {
      setStarving(getStarving() - 1);
    }
  }

  public void produce(Aquarium aquarium) {
    Point posisiKoin = new Point(getPosisi().getAbsis(), getPosisi().getOrdinat());
    Koin koin = new Koin(getLast(), posisiKoin);
    aquarium.addObject(koin);
  }

  public int getProduceTime() {
    return 1;
  }

  public int getMaxProduceTime() {
    return 1;
  }

  public void setProduceTime(int produceTime) {
    // do nothing
  }

  public int getLast() {
    return latestFood;
  }

  public int getId() {
    return id;
  }

  public void setLast(int latestFood) {
    this.latestFood = latestFood;
  }
}